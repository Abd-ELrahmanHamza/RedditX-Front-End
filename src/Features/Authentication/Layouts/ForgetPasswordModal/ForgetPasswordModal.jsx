/* eslint-disable react-hooks/exhaustive-deps */
import {useEffect } from "react";


import FormInput from "Features/Authentication/Components/FormInput/FormInput";

import Button from "Features/Authentication/Components/Button/Button";



import {
  AuthContainer,
  ButtonsContainer,
  ErrorParagraph,
  AuthParagraph,
  AuthHeader,
  Forget,
} from "./ForgetPasswordModal.styled";

const USER_REGEX = /^[A-z][A-z0-9-_]{3,20}$/;

const USER_EMAIL =
  /[a-zA-Z0-9._-]{3,}@[a-zA-Z0-9._-]{3,}[.]{1}[a-zA-Z0-9._-]{2,}/;

/**
 * ForgetPasswordModal component that is used in Login Component
 * @returns {React.Component}  ForgetPasswordModal component that is used in Login Component
 */

const ForgetPasswordModal = ({
  validEmail,
  setValidEmail,
  formFields,
  setFormFields,
  setUserNameScreen,
  validName,
  setValidName,
  passwordScreen,
  setPasswordScreen,
  initialFocus,
  setInitialFocus,
  setErrMsg
}) => {
 
    
  const { userName, email } = formFields;

  

  /**
   * useEffect for userName field to check if the userName that the user entered is valid or not
   */
  useEffect(() => {
    if (userName.length > 0) {
      setInitialFocus(false);
    }
    setValidName(USER_REGEX.test(userName));
  }, [userName]);

  /**
   * useEffect for email field to check if the email that the user entered is valid or not
   */
  useEffect(() => {
    if (email.length > 0) {
      setInitialFocus(false);
      setErrMsg("Not a valid email address");
    } else if (email.length === 0) {
      setErrMsg("Please enter an email address to continue");
      // console.log(errMsg);
    }

    if (userName.length === 0) {
      setErrMsg("Username must be between 3 and 20 characters");
    }
    setValidEmail(USER_EMAIL.test(email));
  }, [email]);

  /**
   * Function to handle the submit of the form of login
   * @param {*} event
   */
  const handleSubmit = async (event) => {
    event.preventDefault();
  };

  /**
   * Function to handle any change on the input field of the login form (check if the userName or the email or the password is valid or not)
   * @param {*} event
   */

  const handleChange = (event) => {
    const { name, value } = event.target;
    if (name === "userName") {
      setValidName(USER_REGEX.test(userName));
    }
    if (name === "email") {
      setValidEmail(USER_EMAIL.test(email));
      console.log(validEmail);
    }

    setFormFields({ ...formFields, [name]: value });
  };

  return (
    <>
      {
        <AuthContainer userNameScreen={!passwordScreen}>
          <AuthHeader>Reset your password</AuthHeader>
          <AuthParagraph>
            Tell us the username and email address associated with your Reddit
            account, and weâ€™ll send you an email with a link to reset your
            password.
          </AuthParagraph>
          <form onSubmit={handleSubmit}>
            <FormInput
              valid={validName}
              initialFocus={initialFocus}
              showIcon={true}
              label="Username"
              type="text"
              onChange={handleChange}
              name="userName"
              value={userName}
              onFocus={() => {
                setInitialFocus(false);
              }}
            />

            {/* Show error message if the userName is not valid and the user made a focus on the it's input field */}
            <ErrorParagraph valid={validName || initialFocus}>
              Username must be between 3 and 20 characters
            </ErrorParagraph>

            <FormInput
              label="Email"
              valid={validEmail}
              initialFocus={initialFocus}
              showIcon={true}
              type="email"
              onChange={handleChange}
              name="email"
              value={email}
            />

            <Forget>
              Forget your{" "}
              <button
                onClick={() => {
                  setInitialFocus(true);
                  setPasswordScreen(false);
                  setUserNameScreen(true);
                  setErrMsg("Not a valid email address");
                }}
              >
                username?
              </button>{" "}
            </Forget>

            <ButtonsContainer>
              {/** Show an enabled button if the email and userName are valid */}
              {validName && validEmail && (
                <Button
                  valid={validName && validEmail}
                  type="button"
                  onClick={() => {}}
                >
                  Reset password
                </Button>
              )}

              {/** Show an disabled button if the email or userName are not valid */}

              {(!validName || !validEmail) && (
                <Button disabled valid={validName && validEmail} type="button">
                  Reset password
                </Button>
              )}
            </ButtonsContainer>

            <Forget>
              Don't have an email or need assistance logging <br /> in?
              <button onClick={() => {}}>Get Help</button>{" "}
            </Forget>
            <br></br>
            <Forget>
              {/* Want to login? */}
              <button
                onClick={() => {
                  setInitialFocus(true);
                  setUserNameScreen(false);
                  setPasswordScreen(false);
                }}
              >
                Log in
              </button>{" "}
            </Forget>
          </form>
        </AuthContainer>
      }
    </>
  );
};

export default ForgetPasswordModal;
