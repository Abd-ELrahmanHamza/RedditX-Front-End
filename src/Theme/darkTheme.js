/**
 * dark theme colors
 *
 * @type {{id:string}}
 */
const darkTheme = {
  // The id of the theme
  id: "dark",
  // The canvas color is the background color of the website
  canvas: "#030303",
  // The colors of the text on the website
  color: {
    // The primary color of the text on the website
    primary: "#d7dadc",
    // The color of the muted text on the website
    muted: "#7c7c7c",
    // The color of the links on the website
    link: "#0079d3",
  },
  // The colors of the backgrounds in the website
  background: {
    primary: "#1a1a1b",
  },
  // The colors of the lines in the website
  lineColor: {
    primary: "#343536",
  },
};

export default darkTheme;
