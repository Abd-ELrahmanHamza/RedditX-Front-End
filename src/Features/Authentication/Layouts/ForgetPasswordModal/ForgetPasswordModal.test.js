import { shallow } from "enzyme";
import { screen, render, fireEvent, waitFor } from "@testing-library/react";
import React from "react";
import { ThemeProvider } from "styled-components";

import darkTheme from "Theme/darkTheme";

import ForgetPasswordModal from "./ForgetPasswordModal";

const defaultFormFields = {
  userName: "",
  password: "",
  email: "",
};

const validEmail = false;
const validName = false;
const passwordScreen = false;
const initialFocus = false;
const initialFocus2 = false;
const setValidEmail = jest.fn();
const setFormFields = jest.fn();
const setUserNameScreen = jest.fn();
const setValidName = jest.fn();
const setPasswordScreen = jest.fn();
const setInitialFocus = jest.fn();
const setInitialFocus2 = jest.fn();
const setErrMsg = jest.fn();

describe("ForgetPasswordModal Layout", () => {
  const wrapper = shallow(
    <ForgetPasswordModal
      validEmail={validEmail}
      validName={validName}
      formFields={defaultFormFields}
      passwordScreen={passwordScreen}
      initialFocus={initialFocus}
      initialFocus2={initialFocus2}
      setValidEmail={setValidEmail}
      setFormFields={setFormFields}
      setUserNameScreen={setUserNameScreen}
      setValidName={setValidName}
      setPasswordScreen={setPasswordScreen}
      setInitialFocus={setInitialFocus}
      setInitialFocus2={setInitialFocus2}
      setErrMsg={setErrMsg}
    />
  );
  expect(wrapper).toMatchSnapshot();

  const userNameInput = wrapper.find('[id="username"]');
  const userNameError = wrapper.find('[id="userNameError"]');
  const forgetUserName = wrapper.find('[id="forgetUserName"]');
  const wantLogIn = wrapper.find('[id="wantLogIn"]');
  const emailInput = wrapper.find('[id="email"]');
  const emailError = wrapper.find('[id="emailError"]');

  it("this is a test for userName field", () => {
    userNameInput.simulate("change", {
      target: { name: "userName", value: "g" },
    });
    expect(userNameError.text()).toEqual(
      "Username must be between 3 and 20 characters"
    );
    expect(setValidName).toBeCalled();
    expect(setInitialFocus).toBeCalled();
    expect(setInitialFocus2).not.toBeCalled();
  });

  it("this is a test for email field", () => {
    emailInput.simulate("change", { target: { name: "email", value: "gdh" } });
    expect(emailError.text()).toEqual("not a valid email address");
    expect(setValidEmail).toBeCalled();
    expect(setInitialFocus2).toBeCalled();
  });

  it("this is a test for forgetUserName link", () => {
    forgetUserName.simulate("click");
    expect(setInitialFocus).toBeCalled();
    expect(setPasswordScreen).toBeCalled();
    expect(setErrMsg).toBeCalled();
    expect(setUserNameScreen).toBeCalled();
  });

  it("this is a test for login link", () => {
    wantLogIn.simulate("click");
    expect(setInitialFocus).toBeCalled();
    expect(setPasswordScreen).toBeCalled();
    expect(setUserNameScreen).toBeCalled();
  });

  test("username-error shouldn't be found", () => {
    render(
      <ThemeProvider theme={darkTheme}>
        <ForgetPasswordModal
          validEmail={validEmail}
          validName={validName}
          formFields={defaultFormFields}
          passwordScreen={passwordScreen}
          initialFocus={initialFocus}
          initialFocus2={initialFocus2}
          setValidEmail={setValidEmail}
          setFormFields={setFormFields}
          setUserNameScreen={setUserNameScreen}
          setValidName={setValidName}
          setPasswordScreen={setPasswordScreen}
          setInitialFocus={setInitialFocus}
          setInitialFocus2={setInitialFocus2}
          setErrMsg={setErrMsg}
        />
      </ThemeProvider>
    );
    expect(screen.queryByTestId("email-error")).not.toBeVisible();
  });

 

  test("username-error should be not found", async () => {
    render(
      <ThemeProvider theme={darkTheme}>
        <ForgetPasswordModal
          validEmail={validEmail}
          validName={validName}
          formFields={defaultFormFields}
          passwordScreen={passwordScreen}
          initialFocus={initialFocus}
          initialFocus2={initialFocus2}
          setValidEmail={setValidEmail}
          setFormFields={setFormFields}
          setUserNameScreen={setUserNameScreen}
          setValidName={setValidName}
          setPasswordScreen={setPasswordScreen}
          setInitialFocus={setInitialFocus}
          setInitialFocus2={setInitialFocus2}
          setErrMsg={setErrMsg}
        />
      </ThemeProvider>
    );
    const userNameInput = screen.getByTestId("username");

    fireEvent.change(userNameInput, {
      target: { value: "Mgfdgs" },
    });

    const userNameError = screen.getByTestId("username-error");

    await waitFor(() => {
      expect(userNameError).not.toBeVisible();
    });
  });

  test("email-input should be found", () => {
    render(
      <ThemeProvider theme={darkTheme}>
        <ForgetPasswordModal
          validEmail={validEmail}
          validName={validName}
          formFields={defaultFormFields}
          passwordScreen={passwordScreen}
          initialFocus={initialFocus}
          initialFocus2={initialFocus2}
          setValidEmail={setValidEmail}
          setFormFields={setFormFields}
          setUserNameScreen={setUserNameScreen}
          setValidName={setValidName}
          setPasswordScreen={setPasswordScreen}
          setInitialFocus={setInitialFocus}
          setInitialFocus2={setInitialFocus2}
          setErrMsg={setErrMsg}
        />
      </ThemeProvider>
    );
    const emailInput = screen.getByTestId("email");
    expect(emailInput).toBeInTheDocument();
  });
});
