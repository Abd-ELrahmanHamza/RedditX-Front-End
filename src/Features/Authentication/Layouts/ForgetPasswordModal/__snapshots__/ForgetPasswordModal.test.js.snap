// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<Fragment>
  <styled.div
    userNameScreen={true}
  >
    <styled.h2>
      Reset your password
    </styled.h2>
    <styled.div>
      Tell us the username and email address associated with your Reddit account, and weâ€™ll send you an email with a link to reset your password.
    </styled.div>
    <form
      onSubmit={[Function]}
    >
      <FormInput
        data-testid="username"
        id="username"
        initialFocus={false}
        label="Username"
        name="userName"
        onChange={[Function]}
        onFocus={[Function]}
        showIcon={true}
        type="text"
        valid={false}
        value=""
      />
      <styled.p
        data-testid="username-error"
        id="userNameError"
        valid={false}
      >
        Username must be between 3 and 20 characters
      </styled.p>
      <FormInput
        data-testid="email"
        id="email"
        initialFocus={false}
        label="Email"
        name="email"
        onChange={[Function]}
        onFocus={[Function]}
        showIcon={true}
        type="email"
        valid={false}
        value=""
      />
      <styled.p
        data-testid="email-error"
        id="emailError"
        valid={false}
      >
        not a valid email address
      </styled.p>
      <styled.span>
        Forget your
         
        <button
          id="forgetUserName"
          onClick={[Function]}
        >
          username?
        </button>
         
      </styled.span>
      <styled.div>
        <Button
          disabled={true}
          onClick={[Function]}
          type="submit"
          valid={false}
        >
          Reset password
        </Button>
      </styled.div>
      <br />
      <styled.span>
        Don't have an email or need assistance logging 
        <br />
         in?
        <button
          onClick={[Function]}
        >
          Get Help
        </button>
         
      </styled.span>
      <br />
      <styled.span>
        <button
          id="wantLogIn"
          onClick={[Function]}
        >
          Log in
        </button>
         
      </styled.span>
    </form>
  </styled.div>
</Fragment>
`;
